manuals/fsapfsinfo.1
.Dd June 26, 2026
.Dt FSAPFSINFO 1
.Os
.Sh NAME
.Nm fsapfsinfo
.Nd determines information about an Apple File System (APFS) container
.Sh SYNOPSIS
.Nm fsapfsinfo
.Op Fl B Ar bodyfile
.Op Fl E Ar identifier
.Op Fl f Ar file_system_index
.Op Fl F Ar path
.Op Fl o Ar offset
.Op Fl p Ar password
.Op Fl r Ar recovery_password
.Op Fl dhHvV
.Ar source
.Sh DESCRIPTION
.Nm fsapfsinfo
is a utility to determine information about an Apple File System (APFS) \
container
.Pp
.Nm fsapfsinfo
is part of the
.Nm libfsapfs
package.
.Nm libfsapfs
is a library to access the Apple File System (APFS) format
.Pp
.Ar source
is the source container.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl B Ar bodyfile
output file system information as a bodyfile
.It Fl d
calculate a MD5 hash of a file entry to include in the bodyfile
.It Fl E Ar identifier
show information about a specific file entry identifier or "all"
.It Fl f Ar file_system_index
show information about a specific file system or "all"
.It Fl F Ar path
show information about a specific file entry path
.It Fl h
shows this help
.It Fl H
shows the file system hierarchy
.It Fl o Ar offset
specify the container offset in bytes
.It Fl p Ar password
specify the password (or passphrase)
.It Fl r Ar recovery_password
specify the recovery password (or passphrase)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fsapfsinfo -o 20480 image.dmg
fsapfsinfo 20260626
.sp
Apple File System (APFS) information:
.sp
Container information:
        Identifier                      : d08a9fa0-d5a5-458b-813e-ebf9bf5d5338
        Number of volumes               : 1
.sp
Volume: 1 information:
        Identifier                      : 458ed10d-8ac3-4af1-8dfd-3954d151a3f3
        Name                            : apfs_test
        Compatible features             : 0x00000002
                (NX_FEATURE_LCFD)

        Incompatible features           : 0x00000001
                (NX_INCOMPAT_VERSION1)

        Read-only compatible features   : 0x00000000

        Number of snapshots             : 0
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr fsapfsmount 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfsapfs/issues
.Sh COPYRIGHT
Copyright (C) 2018-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* rename volume name to volume label ?

* handle transaction identifier mismatch

* fill attribute value_data_file_extents
* complete snapshot support
* extended attribute extent support
  * extent support, add to Python binding
  * move libfsapfs_internal_extended_attribute_get_data_stream to libfsapfs_attributes.[ch]
  * for extended attributes by name look up use native B-tree structures

* handle non-inline attribute symbolic link

* improve format support
** check for more compression methods
** Add resource fork support
** read zlib compressed resource fork data footer
** Check deflate compressed files edge case where first byte is 0x#f
** add support for solaris and bsdos device numbers

* add snapshot support
** add API function to retrieve creation time
** add API function to retrieve change time
** get volume from snapshot?

* B-tree
** Validate (expected) btree level to detect loops - added maximum recursion depth for now
* Validate checksum
** B-tree node
** Container key bag
** Volume key bag
** Reaper
** Object map
** Space manager
* Container superblock read file IO handle add block size support
* use container data block vector in read file IO handle
** Volume superblock
** Container map
** Container key bag?
* Handle file size of 0, mapped file size does not support 0 ?
* Container data handle add decryption support and use it to read container key bag?

library:
* check if data block vector allocated size matches size of file extents
* get_extent function, check inode flags and set extent is sparse flag

* change container read bootstap as advised in documentation
** read backup container superblocks

* API
** Add function to provide volume size
** Add function to provide volume date and time values

* optimization
** read inode and extended attributes at once
** cache uncompressed blocks

* clean up / restructure
** refactor libfsapfs_encryption_aes_key_unwrap to use encryption context?

Python bindings:
* add extents object?

Tools:
* fsapfsinfo tool
** add volume selection via GUID in path
** prompt user for password or recovery password when needed
** print extended attribute values
** handle \r in filename
** print hardlinks for -E option?
** "-E all" instead of using next_object_identifier walk leaf nodes

* fsapfsmount tool
** prompt user for password or recovery password when needed
** make symbolic links an absolute path which accounts for mount point
** add owner and group identifier support?
** add support for all volumes
** add volume selection support
** does fuse override file types such as socket?
** dokan add symbolic link support

Tests:
* add tests for container_key_bag
* add tests for volume_key_bag
* add tests for key_encrypted_key

20180907
* see `git log' for more recent change log
* Initial version based on libfsrefs and libfsntfs



fsapfstools/fsapfsinfo.c
/*
 * Shows information obtained from an Apple File System (APFS) container.
 *
 * Copyright (C) 2018-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fsapfstools_getopt.h"
#include "fsapfstools_libcerror.h"
#include "fsapfstools_libclocale.h"
#include "fsapfstools_libcnotify.h"
#include "fsapfstools_libfsapfs.h"
#include "fsapfstools_output.h"
#include "fsapfstools_signal.h"
#include "fsapfstools_unused.h"
#include "info_handle.h"

enum FSAPFSINFO_MODES
{
	FSAPFSINFO_MODE_CONTAINER,
	FSAPFSINFO_MODE_FILE_ENTRIES,
	FSAPFSINFO_MODE_FILE_ENTRY_BY_IDENTIFIER,
	FSAPFSINFO_MODE_FILE_ENTRY_BY_PATH,
	FSAPFSINFO_MODE_FILE_SYSTEM_HIERARCHY
};

info_handle_t *fsapfsinfo_info_handle = NULL;
int fsapfsinfo_abort                  = 0;

/* Signal handler for fsapfsinfo
 */
void fsapfsinfo_signal_handler(
      fsapfstools_signal_t signal FSAPFSTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fsapfsinfo_signal_handler";

	FSAPFSTOOLS_UNREFERENCED_PARAMETER( signal )

	fsapfsinfo_abort = 1;

	if( fsapfsinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     fsapfsinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fsapfsinfo to determine information about an Apple File System (APFS) container.";

	fsapfstools_option_t options[ ] = {
		{ 'B', "bodyfile", "output file system information as a bodyfile" },
		{ 'd', NULL, "calculate a MD5 hash of a file entry to include in the bodyfile" },
		{ 'E', "identifier", "show information about a specific file entry identifier or \"all\"" },
		{ 'f', "file_system_index", "show information about a specific file system or \"all\"" },
		{ 'F', "path", "show information about a specific file entry path" },
		{ 'h', NULL, "shows this help" },
		{ 'H', NULL, "shows the file system hierarchy" },
		{ 'o', "offset", "specify the container offset in bytes" },
		{ 'p', "password", "specify the password (or passphrase)" },
		{ 'r', "recovery_password", "specify the recovery password (or passphrase)" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source container" },
	};
	system_character_t options_string[ 32 ];

	libfsapfs_error_t *error                         = NULL;
	system_character_t *option_bodyfile              = NULL;
	system_character_t *option_file_entry_identifier = NULL;
	system_character_t *option_file_entry_path       = NULL;
	system_character_t *option_file_system_index     = NULL;
	system_character_t *option_password              = NULL;
	system_character_t *option_recovery_password     = NULL;
	system_character_t *option_volume_offset         = NULL;
	system_character_t *source                       = NULL;
	char *program                                    = "fsapfsinfo";
	system_integer_t option                          = 0;
	size_t string_length                             = 0;
	uint64_t file_entry_identifier                   = 0;
	uint8_t calculate_md5                            = 0;
	int number_of_options                            = (int) ( sizeof( options ) / sizeof( fsapfstools_option_t ) );
	int option_mode                                  = FSAPFSINFO_MODE_CONTAINER;
	int verbose                                      = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fsapfstools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fsapfstools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fsapfstools_output_version_fprint(
	 stdout,
	 program );

	if( fsapfstools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fsapfstools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fsapfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'B':
				option_bodyfile = optarg;

				break;

			case (system_integer_t) 'd':
				calculate_md5 = 1;

				break;

			case (system_integer_t) 'E':
				option_mode                  = FSAPFSINFO_MODE_FILE_ENTRY_BY_IDENTIFIER;
				option_file_entry_identifier = optarg;

				break;

			case (system_integer_t) 'f':
				option_file_system_index = optarg;

				break;

			case (system_integer_t) 'F':
				option_mode            = FSAPFSINFO_MODE_FILE_ENTRY_BY_PATH;
				option_file_entry_path = optarg;

				break;

			case (system_integer_t) 'h':
				fsapfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'H':
				option_mode = FSAPFSINFO_MODE_FILE_SYSTEM_HIERARCHY;

				break;

			case (system_integer_t) 'o':
				option_volume_offset = optarg;

				break;

			case (system_integer_t) 'p':
				option_password = optarg;

				break;

			case (system_integer_t) 'r':
				option_recovery_password = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fsapfstools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source container.\n" );

		fsapfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libfsapfs_notify_set_stream(
	 stderr,
	 NULL );
	libfsapfs_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &fsapfsinfo_info_handle,
	     calculate_md5,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( fsapfsinfo_info_handle != NULL );
#endif
	if( option_bodyfile != NULL )
	{
		if( info_handle_set_bodyfile(
		     fsapfsinfo_info_handle,
		     option_bodyfile,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set bodyfile.\n" );

			goto on_error;
		}
	}
	if( option_file_system_index != NULL )
	{
		if( info_handle_set_file_system_index(
		     fsapfsinfo_info_handle,
		     option_file_system_index,
		     &error ) != 1 )
		{
			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );

			fprintf(
			 stderr,
			 "Unsupported file system index defaulting to: all.\n" );
		}
	}
	if( option_password != NULL )
	{
		if( info_handle_set_password(
		     fsapfsinfo_info_handle,
		     option_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set password.\n" );

			goto on_error;
		}
	}
	if( option_recovery_password != NULL )
	{
		if( info_handle_set_recovery_password(
		     fsapfsinfo_info_handle,
		     option_recovery_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set recovery password.\n" );

			goto on_error;
		}
	}
	if( option_volume_offset != NULL )
	{
		if( info_handle_set_volume_offset(
		     fsapfsinfo_info_handle,
		     option_volume_offset,
		     &error ) != 1 )
		{
			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );

			fprintf(
			 stderr,
			 "Unsupported volume offset defaulting to: %" PRIi64 ".\n",
			 fsapfsinfo_info_handle->volume_offset );
		}
	}
	if( info_handle_open_input(
	     fsapfsinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( option_mode == FSAPFSINFO_MODE_FILE_ENTRY_BY_IDENTIFIER )
	{
		if( option_file_entry_identifier == NULL )
		{
			fprintf(
			 stderr,
			 "Mising file entry identifier string.\n" );

			goto on_error;
		}
		string_length = system_string_length(
				 option_file_entry_identifier );

		if( ( string_length == 3 )
		 && ( system_string_compare(
		       option_file_entry_identifier,
		       _SYSTEM_STRING( "all" ),
		       3 ) == 0 ) )
		{
			option_mode = FSAPFSINFO_MODE_FILE_ENTRIES;
		}
		else if( info_handle_system_string_copy_from_64_bit_in_decimal(
		          option_file_entry_identifier,
		          string_length + 1,
		          &file_entry_identifier,
		          &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to copy file entry identifier string to 64-bit decimal.\n" );

			goto on_error;
		}
		else if( file_entry_identifier > (uint64_t) 0x0fffffffffffffffUL )
		{
			fprintf(
			 stderr,
			 "Invalid file entry identifier value out of bounds." );

			goto on_error;
		}
	}
	switch( option_mode )
	{
		case FSAPFSINFO_MODE_FILE_ENTRIES:
			if( info_handle_file_entries_fprint(
			     fsapfsinfo_info_handle,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print file entries.\n" );

				goto on_error;
			}
			break;

		case FSAPFSINFO_MODE_FILE_ENTRY_BY_IDENTIFIER:
			if( info_handle_file_entry_fprint_by_identifier(
			     fsapfsinfo_info_handle,
			     file_entry_identifier,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print file entry: %" PRIu64 ".\n",
				 file_entry_identifier );

				goto on_error;
			}
			break;

		case FSAPFSINFO_MODE_FILE_ENTRY_BY_PATH:
			if( info_handle_file_entry_fprint_by_path(
			     fsapfsinfo_info_handle,
			     option_file_entry_path,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print file entry information.\n" );

				goto on_error;
			}
			break;

		case FSAPFSINFO_MODE_FILE_SYSTEM_HIERARCHY:
			if( info_handle_file_system_hierarchy_fprint(
			     fsapfsinfo_info_handle,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print file system hierarchy.\n" );

				goto on_error;
			}
			break;

		case FSAPFSINFO_MODE_CONTAINER:
		default:
			if( info_handle_container_fprint(
			     fsapfsinfo_info_handle,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print container information.\n" );

				goto on_error;
			}
			break;
	}
	if( info_handle_close_input(
	     fsapfsinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &fsapfsinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( fsapfsinfo_info_handle != NULL )
	{
		info_handle_free(
		 &fsapfsinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

